src/ParsingMiddleware.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace Amp\Http\Server\FormParser;
+
+use Amp\Http\Server\ErrorHandler;
+use Amp\Http\Server\Middleware;
+use Amp\Http\Server\Request;
+use Amp\Http\Server\RequestHandler;
+use Amp\Http\Server\Server;
+use Amp\Http\Server\ServerObserver;
+use Amp\Http\Status;
+use Amp\Promise;
+use Amp\Success;
+use function Amp\call;
+
+class ParsingMiddleware implements Middleware, ServerObserver
+{
+    /** @var BufferingParser */
+    private $parser;
+
+    /** @var ErrorHandler */
+    private $errorHandler;
+
+    public function __construct(int $fieldCountLimit = null)
+    {
+        $this->parser = new BufferingParser($fieldCountLimit);
+    }
+
+    public function handleRequest(Request $request, RequestHandler $requestHandler): Promise
+    {
+        return call(function () use ($request, $requestHandler) {
+            try {
+                $request->setAttribute(self::class, yield $this->parser->parseForm($request));
+            } catch (ParseException $exception) {
+                return yield $this->errorHandler->handleError(Status::PAYLOAD_TOO_LARGE, null, $request);
```

Comment:
Is a `ParseException` only thrown on too large forms? I don't think so.

Reply:
It's thrown if there are too many fields, that's all in `BufferingParser`. Maybe 413 isn't the right error code, perhaps just 400? We may want to reconsider using `explode()` to parse, since that creates the enormous array we're trying to avoid with the limit.

Reply:
I guess it should be just 400, yes.